### PR TITLE
add workflows release

### LINKED ISSUE #81: Publish new release by Github Action
### What would you like to be added?

Add a new Github Action for new releases. Like this: https://github.com/kuasar-io/kuasar/releases/tag/v0.1.0

The fist tar file `kuasar-vx.y.z-linux-amd64.tar.gz` compresses this binaries:
+ `vmm-sandboxer`(for Cloud Hypervisor) and `config_clh.toml`
+ `kuasar.image` and `vmlinux.bin` for Cloud Hypervisor
+ `wasm-sandboxer`(for wasmEdge)
+ `quark-sandboxer`
+ `containerd` binary and `config.toml` of Kuasar community
+ `shim` folder with two shim binaries

The second tar file `kuasar-vx.y.z-vendor.tar.gz` compresses the source code with vendor code.

The x.y.z represents the tag or version number of this release.

### Why is this needed?

Currently, Kuasar releases are manually executed and should be improved with automatic pipeline.

### PR DESCRIPTION
**Pull request by dierbei**
Fix #81

### CODE REVIEW

**Comment by xiaods** [comment]
why not merged it?

**Comment by xiaods** [comment]
suggest squash commit to one commit. give more context on your commit log.

**Comment by Burning1020** [comment]
> suggest squash commit to one commit. give more context on your commit log.

@dierbei Yes, please squash your commits, thanks.

**Comment by dierbei** [comment]
Thanks for the suggestion. Wait a minute, I still need to make some adjustments.

**Comment by dierbei** [comment]
@Burning1020 @xiaods Already rebase. please help me to review, thanks.

File: .github/workflows/release.yaml [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+            address = "/run/vmm-sandboxer.sock"
+
+          [proxy_plugins.quark]
+            type = "sandbox"
+            address = "/run/quark-sandboxer.sock"
+
+          [proxy_plugins.wasm]
+            type = "sandbox"
+            address = "/run/wasm-sandboxer.sock"
+          EOL
```
Comment (Burning1020) [inline .github/workflows/release.yaml:151]: How about touch a containerd file and copy it in `Makefile`

File: .github/workflows/release.yaml [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+          name: ${{ matrix.directories }}
+          path: shim/bin/
+        
+  kuasar-image:
+    runs-on: ubuntu-latest
+    timeout-minutes: 30
+    steps:
+      - uses: actions/checkout@v3
+      - name: Build kuasar.img
+        run: RUNTIME=docker make bin/kuasar.img
```
Comment (Burning1020) [inline .github/workflows/release.yaml:83]: Add `sudo` to here, refer to: https://github.com/kuasar-io/kuasar/blob/main/.github/workflows/ci.yml#L76

File: .github/workflows/release.yaml [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+          echo "RELEASE_VERSION=${GITHUB_REF:10}" >> $GITHUB_ENV
+          echo "ARCHIVE_NAME=kuasar-${GITHUB_REF:10}-linux-amd64.tar.gz" >> $GITHUB_ENV
+          echo "VENDOR_NAME=kuasar-${GITHUB_REF:10}-vendor.tar.gz" >> $GITHUB_ENV
+      - name: Package vendor
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          gh release create ${{ env.RELEASE_VERSION }} --generate-notes
+          mkdir ../temp
+          cp -r ./* ../temp/
```
Comment (Burning1020) [inline .github/workflows/release.yaml:181]: We should provide a package that contains all vendor source code, so we need use `cargo vendor`  to generate all vendor codes under every cargo folders.

File: .github/workflows/release.yaml [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+          path: ${{ env.BIN_DIR }}
+
+  release:
+    permissions:
+      contents: write
+    needs:
+      - buildSandboxer
+      - buildShim
+      - kuasar-image
+      - vmlinux-bin
```
Comment (Burning1020) [inline .github/workflows/release.yaml:165]: We have `vmm quark wasm` target in Makefile. Maybe `make vmm`, `make quark` and `make wasm` could make it simple.

File: scripts/build/gen-containerd-config.sh [hunk 5]
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash
+
+sudo curl -LJO https://github.com/containerd/containerd/releases/download/v1.7.0/containerd-1.7.0-linux-amd64.tar.gz
+sudo mkdir bin && sudo tar -C bin -xzvf containerd-1.7.0-linux-amd64.tar.gz
```
Comment (Burning1020) [inline scripts/build/gen-containerd-config.sh:4]:
```shell
git clone https://github.com/kuasar-io/containerd.git
make -C containerd bin/containerd
mkdir bin && mv containerd/bin/containerd bin
```

File: scripts/build/gen-containerd-config.sh [hunk 6]
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash
+
```
Comment (Burning1020) [inline scripts/build/gen-containerd-config.sh:2]: Add filehead and `set -e`:https://github.com/kuasar-io/kuasar/blob/main/examples/run_example_container.sh#L2-L16

File: scripts/build/gen-containerd-config.sh [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+sudo curl -LJO https://github.com/containerd/containerd/releases/download/v1.7.0/containerd-1.7.0-linux-amd64.tar.gz
+sudo mkdir bin && sudo tar -C bin -xzvf containerd-1.7.0-linux-amd64.tar.gz
+
+sudo tee bin/config.toml > /dev/null <<EOF
```
Comment (Burning1020) [inline scripts/build/gen-containerd-config.sh:6]: Is it necessary to add `sudo` in front of every commands? I mean it is possiable to add it in `job.step.run`, https://github.com/kuasar-io/kuasar/pull/95#discussion_r1337032458

**Review by Burning1020 — CHANGES_REQUESTED** [review]

File: scripts/build/download-containerd.sh [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -e
+
+git clone https://github.com/kuasar-io/containerd.git 
+mkdir bin && make -C containerd bin/containerd && mv containerd/bin/containerd bin
```
Comment (Burning1020) [inline scripts/build/download-containerd.sh:19]: Append `rm -rf containerd`, we need to cleanup.

File: .github/workflows/release.yaml [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+            bin/containerd-shim-kuasar-vmm-v2 
+            bin/containerd-shim-kuasar-wasm-v2
+
+  containerd:
+    runs-on: ubuntu-latest
+    timeout-minutes: 30
+    steps:
+      - uses: actions/checkout@v3
+      - name: Build Containerd
+        run: sudo -E ./scripts/build/download-containerd.sh
```
Comment (Burning1020) [inline .github/workflows/release.yaml:91]: `build-containerd.sh`, not `donwload-containerd.sh`

**Review by Burning1020 — APPROVED** [review]
LGTM. Well done.

**Pull request merged**